### PR TITLE
Stop cause and ccctf

### PR DESCRIPTION
**Pull request by rovere**
## Stop reason for a track/trajectory

Each TrajectoryFilter has been equipped in such a way to
inject into the Trajectory or TempTrajectory the reason that
stopped the propagation of the trajectory during the pattern
recognition. This work is in preparation of future work that
will propagate this information down the chain directly into
a reco::TrackBase, so that more detailed studies can be made
on the reasons of the stop.
## CCC as a Trajectory Filter

The CCC has been implemented as a TrajectoryFilter, in order to give
more flexibility to the user in selecting possible bad (from the
Cluster Charge point of view) hits. The final trajectory, if discarded due 
to eccessive number of bad CCC hits, will anyhow contain the
would-have-been removed 2 hits, since they passed all the other estimators. 
The user has the possibility to disentangle lost hits from bad for CCC
and apply different criteria with different thresholds during Patter
Recognition and at filtering level.

**NOTE** 
The CCC as Trajectory filter has been turned off, so that we should expect no changes and have this PR integrated quickly. The usage and tuning of this new features will be done on top of this. Private results are anyway extremely encouraging (w.r.t to the hit collection efficiency).

### CODE REVIEW

**Comment by VinInn** [comment]
@cmsbuild, please test

File: TrackingTools/TrajectoryFiltering/interface/MaxCCCLostHitsTrajectoryFilter.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+public:
+
+  explicit MaxCCCLostHitsTrajectoryFilter (int maxCCCHits=0,
+                                           float CCC_value=0) :
+      theMaxCCCLostHits_(maxCCCHits),
+      minGoodStripCharge_(CCC_value) {}
+
+  explicit MaxCCCLostHitsTrajectoryFilter (
+      const edm::ParameterSet & pset, edm::ConsumesCollector& iC) :
+      theMaxCCCLostHits_(pset.getParameter<int>("maxCCCLostHits")),
```
Comment (VinInn) [inline TrackingTools/TrajectoryFiltering/interface/MaxCCCLostHitsTrajectoryFilter.h]: protect these...

**Comment by slava77** [comment]
please test (compile and run the short matrix) before updating the PR or requesting another jenkins tests.
There is a limited amount of resources and space for the tests and it can interfere with other activities.

**Comment by VinInn** [comment]
@cmsbuild, please test

File: TrackingTools/PatternTools/interface/TempTrajectory.h [hunk 2]
```diff
@@ -206,6 +229,11 @@ class TempTrajectory
    */
   int lostHits() const { return theNumberOfLostHits;}
 
+  /** Number of hits that are not compatible with the CCC used during
+   *  patter recognition. Used mainly as a criteria for abandoning a
+   *  trajectory candidate during trajectory building.
+   */
+  int CCCBadHits() const { return theNumberOfCCCBadHits_;}
```
Comment (slava77) [inline TrackingTools/PatternTools/interface/TempTrajectory.h]: please keep to CMS style: member methods start with a lower case letter

File: TrackingTools/PatternTools/interface/Trajectory.h [hunk 3]
```diff
@@ -240,6 +302,12 @@ class Trajectory
 
   int lostHits() const { return theNumberOfLostHits;}
 
+  /** Number of hits that are not compatible with the CCC used during
+   *  patter recognition. Used mainly as a criteria for abandoning a
+   *  trajectory candidate during trajectory building.
+   */
+  int CCCBadHits() const { return theNumberOfCCCBadHits_;}
```
Comment (slava77) [inline TrackingTools/PatternTools/interface/Trajectory.h]: as in TempTrajectory:
start with a lower case
``` cccBadHits``` is just as good
... in general, I find it a bad idea to define a category based on one particular cut. If we redo this cut to do something else (not just cluster charge), the name will become inappropriate.

File: TrackingTools/PatternTools/interface/Trajectory.h [hunk 4]
```diff
@@ -326,6 +394,8 @@ class Trajectory
   /// two layers crossed by the trajectory
    float dPhiCacheForLoopersReconstruction() const { return theDPhiCache;}
 
+   float CCCThreshold() const {return theCCCThreshold_;}
```
Comment (slava77) [inline TrackingTools/PatternTools/interface/Trajectory.h]: ccc

**Comment by slava77** [comment]
@rovere 
please update the method names, after that this PR can be signed off quickly.
Thanks.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/12576	265e83b
- code changes are in line with the description
- previous jenkins tests OK: passed and comparisons showed no differece; the last change is trivial (jenkins compiled OK)

**Comment by davidlange6** [comment]
+1

**Pull request merged**